### PR TITLE
remove wait queue from ruby thread pool to avoid possibility of deadlock

### PR DESCRIPTION
**Pull request by apolcyn**
fixes https://github.com/grpc/grpc/issues/8603
similar results on local benchmarks btw

### CODE REVIEW

File: src/ruby/spec/generic/rpc_server_pool_spec.rb [hunk 1]
```diff
@@ -44,32 +46,34 @@
...
-      3.times do
+      5.times do
+        expect(p.ready_for_work?).to be(true)
         p.schedule(&job)
       end
-      p.start
-      sleep 2
-      expect(p.jobs_waiting).to be(0)
+      expect(p.ready_for_work?).to be(false)
+      sleep 2 # give the pool time do get at least one task done
```
Comment (vjpai) [inline src/ruby/spec/generic/rpc_server_pool_spec.rb]: sleep 3 or even 5 just to be on the safe side?

File: src/ruby/spec/generic/rpc_server_pool_spec.rb [hunk 2]
```diff
@@ -44,32 +46,34 @@
...
+      job = proc { sleep(3) } # sleep so workers busy when done scheduling
+      5.times do
+        expect(p.ready_for_work?).to be(true)
         p.schedule(&job)
-        expect(p.jobs_waiting).to be(i + 1)
       end
+      expect(p.ready_for_work?).to be(false)
     end
 
     it 'it decreases as jobs are run' do
```
Comment (vjpai) [inline src/ruby/spec/generic/rpc_server_pool_spec.rb]: This description doesn't sounds good. How about 'it becomes ready again after jobs complete'

File: src/ruby/spec/generic/rpc_server_pool_spec.rb [hunk 3]
```diff
@@ -90,6 +94,18 @@
       expect(q.pop).to be(o)
       p.stop
     end
+
+    it 'it throws an error if all opf the workers have tasks to do' do
```
Comment (vjpai) [inline src/ruby/spec/generic/rpc_server_pool_spec.rb]: of

File: src/ruby/spec/generic/rpc_server_pool_spec.rb [hunk 4]
```diff
@@ -114,17 +130,17 @@
 
   describe '#start' do
     it 'runs pre-scheduled jobs' do
```
Comment (vjpai) [inline src/ruby/spec/generic/rpc_server_pool_spec.rb]: But this test is no longer about pre-scheduled jobs, correct? In fact, the concept is now meaningless (and maybe always was). How about just dropping this test or changing it to something else

File: src/ruby/spec/generic/rpc_server_pool_spec.rb [hunk 5]
```diff
@@ -55,15 +55,15 @@
     it 'it stops being ready after all workers jobs waiting or running' do
       p = Pool.new(5)
       p.start
-      job = proc { sleep(3) } # sleep so workers busy when done scheduling
+      job = proc { sleep(5) } # sleep so workers busy when done scheduling
```
Comment (vjpai) [inline src/ruby/spec/generic/rpc_server_pool_spec.rb]: Why change this one?

Comment (apolcyn) [inline src/ruby/spec/generic/rpc_server_pool_spec.rb]: I just changed that one to 5 seconds too after changing the other one to 5 seconds. But I guess this is probably an unnecessary wait. I can change this back to 3

**Comment by vjpai** [comment]
Remaining test fails are all non-related.

**Comment by rafaelsales** [comment]
Having a waiting queue was a huge simplification of clients - imagine if web servers did not support waiting queue!
Is there a way to implement the implement a waiting queue with a sort of middleware and plug in RpcServer?

**Comment by apolcyn** [comment]
@rafaelsales I'm unclear sorry, but is this an issue for you? What's your use case?

**Comment by apolcyn** [comment]
@rafaelsales looking at this PR again after a long time. FTR I think that the problem you're describing would be most easily fixed by setting the `pool_size` parameter to increase size of the server's thread pool, as is done in [this test server](https://github.com/grpc/grpc/blob/master/src/ruby/qps/server.rb#L53). I'd have a hard time imaging a scenario that relies specifically on the "queue" behavior for waiting-to-be-serviced calls and that can't be fixed by increasing the server's concurrent RPC capacity, but it would be interesting to know if there was one.

**Pull request merged**